From 722e5420ebf9fd46eed93f21f4af9e4bcbbbc502 Mon Sep 17 00:00:00 2001
From: Atchaya R <quic_ratchaya@quicinc.com>
Date: Thu, 27 Jun 2024 00:28:39 +0530
Subject: [PATCH] wifi: sbin/wifi script

/sbin/wifi: Allow pre/post configuration hook in wifi script
Fix wifi_detect if explicit driver type is provided
Enable 11ad hw_mode support
Enable provision to unload/load wifi modules
Fix enabling provision to unload,load wifi modules
Fix wifi_load if explicit driver type is provided
Add support for modules unload
Disable wifi interfaces during module unload
Adding support to handle traps in the wifi script
"wifi recover" command supported to recover VAPs.
Adding support for WEP152 in wifi script
Add radio specific recover support in wifi script
"Wifi detect is deprecated. Use wifi config instead"
Modify script to do VAP level wifi down/up
Add support to seperate out create and start vaps
Add support for MLD  option
wifi down/up for multi radio, multi VAPs
Fix post_driver call to send parameters as array
skip wifi bringup if /proc/cmdline has skip_wifi parameter
Add a new option in wifi script to carry out MLO operations
Align to 12.x features and network standby

Signed-off-by: Atchaya R <quic_ratchaya@quicinc.com>
Change-Id: If1c55206d0dcf40151e892b455a8b22e142735ab
(cherry picked from commit 1d24b66e112f250b5febb05fd3bd40547e0d4c30)
---
 .../config/wifi-scripts/files/sbin/wifi       | 128 ++++++++++++++++--
 1 file changed, 113 insertions(+), 15 deletions(-)

diff --git a/package/network/config/wifi-scripts/files/sbin/wifi b/package/network/config/wifi-scripts/files/sbin/wifi
index f937dba7e6d9..3ef603b17618 100755
--- a/package/network/config/wifi-scripts/files/sbin/wifi
+++ b/package/network/config/wifi-scripts/files/sbin/wifi
@@ -78,7 +78,8 @@ prepare_key_wep() {
 
 	echo -n "$key" | grep -qE "[^a-fA-F0-9]" && hex=0
 	[ "${#key}" -eq 10 -a $hex -eq 1 ] || \
-	[ "${#key}" -eq 26 -a $hex -eq 1 ] || {
+	[ "${#key}" -eq 26 -a $hex -eq 1 ] || \
+	[ "${#key}" -eq 32 -a $hex -eq 1 ] || {
 		[ "${key:0:2}" = "s:" ] && key="${key#s:}"
 		key="$(echo -n "$key" | hexdump -ve '1/1 "%02x" ""')"
 	}
@@ -93,6 +94,7 @@ wifi_fixup_hwmode() {
 	config_get channel "$device" channel
 	config_get hwmode "$device" hwmode
 	case "$hwmode" in
+		11ad) hwmode=ad;;
 		11bg) hwmode=bg;;
 		11a) hwmode=a;;
 		11ad) hwmode=ad;;
@@ -123,6 +125,12 @@ wifi_fixup_hwmode() {
 }
 
 _wifi_updown() {
+	for driver in ${DRIVERS}; do (
+		if eval "type pre_${driver}" 2>/dev/null >/dev/null; then
+			eval "pre_${driver} ${1} '$2' '$3' '$4' '$5'"
+		fi
+	); done
+
 	for device in ${2:-$DEVICES}; do (
 		config_get disabled "$device" disabled
 		[ "$disabled" = "1" ] && {
@@ -132,19 +140,26 @@ _wifi_updown() {
 		config_get iftype "$device" type
 		if eval "type ${1}_$iftype" 2>/dev/null >/dev/null; then
 			eval "scan_$iftype '$device'"
-			eval "${1}_$iftype '$device'" || echo "$device($iftype): ${1} failed"
+			eval "${1}_$iftype '$device' '$3'" || echo "$device($iftype): ${1} failed"
 		elif [ ! -f /lib/netifd/wireless/$iftype.sh ]; then
 			echo "$device($iftype): Interface type not supported"
 		fi
 	); done
+
+	for driver in ${DRIVERS}; do (
+		if eval "type post_${driver}" 2>/dev/null >/dev/null; then
+			eval "post_${driver} ${1} '$2' '$3'"
+		fi
+	); done
 }
 
 wifi_updown() {
+	pre_wifi_updown
 	cmd=down
 	[ enable = "$1" ] && {
-		_wifi_updown disable "$2"
+		_wifi_updown disable "$2" "$3"
 		ubus_wifi_cmd "$cmd" "$2"
-		ubus call network reload
+		[ "$DISABLE_NET_RELOAD" == "1" ] || ubus call network reload
 		scan_wifi
 		cmd=up
 	}
@@ -155,17 +170,25 @@ wifi_updown() {
 	}
 	ubus_wifi_cmd "$cmd" "$2"
 	_wifi_updown "$@"
+	post_wifi_updown
+}
+
+wifi_recover() {
+	_wifi_updown disable_recover "$1"
+	_wifi_updown enable_recover "$1"
 }
 
 wifi_reload_legacy() {
-	_wifi_updown "disable" "$1"
+	pre_wifi_reload_legacy
+	_wifi_updown "disable" "$1" "$2" "$3" "$4"
 	scan_wifi
-	_wifi_updown "enable" "$1"
+	_wifi_updown "enable" "$1" "$2" "$3" "$4"
+	post_wifi_reload_legacy
 }
 
 wifi_reload() {
 	ubus call network reload
-	wifi_reload_legacy
+	wifi_reload_legacy "" "$2"
 }
 
 wifi_detect_notice() {
@@ -175,18 +198,56 @@ wifi_detect_notice() {
 }
 
 wifi_config() {
-	[ -e /tmp/.config_pending ] && return
-	ucode /usr/share/hostap/wifi-detect.uc
+	pre_wifi_config
 	[ ! -f /etc/config/wireless ] && touch /etc/config/wireless
 	ucode /lib/wifi/mac80211.uc | uci -q batch
 
 	for driver in $DRIVERS; do (
 		if eval "type detect_$driver" 2>/dev/null >/dev/null; then
-			eval "detect_$driver" || echo "$driver: Detect failed" >&2
+			eval "detect_$driver" "$1" || echo "$driver: Detect failed" >&2
 		else
 			echo "$driver: Hardware detection not supported" >&2
 		fi
 	); done
+	post_wifi_config
+}
+
+wifi_unload() {
+	pre_wifi_unload
+	_wifi_updown "disable" "$1"
+	for driver in ${1:-$DRIVERS}; do (
+		if eval "type unload_$driver" 2>/dev/null >/dev/null; then
+			eval "unload_$driver" || echo "$driver: unload failed" >&2
+		else
+			echo "$driver: wifi unload not supported" >&2
+		fi
+	); done
+	post_wifi_unload
+}
+
+_wifi_unload() {
+	_wifi_updown "disable" "$1"
+	for driver in ${1:-$DRIVERS}; do (
+		if eval "type unload_$driver" 2>/dev/null >/dev/null; then
+			eval "unload_$driver" || echo "$driver: unload failed" >&2
+		else
+			echo "$driver: wifi unload not supported" >&2
+		fi
+	); done
+}
+
+wifi_load() {
+	pre_wifi_load
+	_wifi_unload
+	for driver in ${1:-$DRIVERS}; do (
+		if eval "type load_$driver" 2>/dev/null >/dev/null; then
+			eval "load_$driver" "$2" || echo "$driver: load failed" >&2
+		else
+			echo "$driver: wifi load not supported" >&2
+		fi
+	); done
+	_wifi_updown "enable" "$1" "$2"
+	post_wifi_load
 }
 
 start_net() {(
@@ -227,6 +288,7 @@ set_wifi_down() {
 scan_wifi() {
 	local cfgfile="$1"
 	DEVICES=
+	MLD_DEVICES=
 	config_cb() {
 		local type="$1"
 		local section="$2"
@@ -249,26 +311,62 @@ scan_wifi() {
 				append vifs "$CONFIG_SECTION"
 				config_set "$device" vifs "$vifs"
 			;;
+			wifi-mld)
+				if ! list_contains MLD_DEVICES "$CONFIG_SECTION" ; then
+					append MLD_DEVICES "$CONFIG_SECTION"
+				fi
+			;;
 		esac
 	}
 	config_load "${cfgfile:-wireless}"
 }
 
+wifi_trap() {
+	for driver in ${1:-$DRIVERS}; do (
+		if eval "type trap_$driver" 2>/dev/null >/dev/null; then
+			eval "trap_$driver" >&2
+		fi
+	); done
+}
+
 DEVICES=
+MLD_DEVICES=
 DRIVERS=
 include /lib/wifi
 scan_wifi
 
+trap 'wifi_trap; exit' INT TERM ABRT QUIT ALRM
+
+if grep -q 'skip_wifi' /proc/cmdline; then
+	echo "skipping wifi bringup" > /dev/console
+	exit 0
+fi
+
+DISABLE_NET_RELOAD=0
+type check_network_reload &>/dev/null && {
+        check_network_reload
+        DISABLE_NET_RELOAD=$?
+        echo "Disable Network Reload $DISABLE_NET_RELOAD" > /dev/console
+}
+
 case "$1" in
-	down) wifi_updown "disable" "$2";;
+	down) wifi_updown "disable" "$2" "$3";;
 	detect) wifi_detect_notice ;;
-	config) wifi_config ;;
+	config) wifi_config "$2";;
 	status) ubus_wifi_cmd "status" "$2";;
 	isup) wifi_isup "$2"; exit $?;;
-	reload) wifi_reload "$2";;
-	reload_legacy) wifi_reload_legacy "$2";;
+	reload) wifi_reload "$2" "wifi_reload";;
+	reload_legacy) wifi_reload_legacy "$2" "" "$1" "$3";;
+	event_reload_legacy) event_wifi_updown "enable" "$2";;
+	load) wifi_load "$2" "wifi_load";;
+	unload) wifi_unload "$2";;
+	recover) wifi_recover "$2";;
+	mlo) mlo "$@";;
+	standby) standby_wifi_radio_updown "$@" ;;
+        multi_up|\
+        multi_down) [ "$DISABLE_NET_RELOAD" == "1" ] || multi_radio_wifi_updown "$@";;
 	--help|help) usage;;
 	reconf) wifi_updown "reconf" "$2";;
-	''|up) wifi_updown "enable" "$2";;
+	''|up) [ "$DISABLE_NET_RELOAD" == "1" ] || wifi_updown "enable" "$2" "$3" "wifi_up";;
 	*) usage; exit 1;;
 esac
-- 
2.34.1

